workshop/session_3/11_biopython_proteins.ipynb

```diff
@@ -0,0 +1,957 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": false
+   },
+   "source": [
+    "PDB files contain the hierarchical representation consisting of:\n",
+    " - `structure` at the top level\n",
+    " - `model` nested underneath\n",
+    " - `chain` - in this case there are 2\n",
+    " - `residue` - belonging to a chain - a particular aminoacid\n",
+    " - `atom` - which contains the coordinates\n",
+    " \n",
+    " More information on the PDB structure and representation can be [found here](https://pdb101.rcsb.org/learn/guide-to-understanding-pdb-data/introduction).\n",
+    " \n",
+    " Let's traverse the parsed file and calculate some statistics. We will be interested in rhe total TRP count and the number of carbon atoms.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 175,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Total number of TRP aminoacids in the structure: 22\n",
+      "Total number of carbon atoms in the structure:   6347\n"
+     ]
+    }
+   ],
+   "source": [
+    "total_TRP_residues = 0\n",
+    "total_carbon_atoms = 0\n",
+    "\n",
+    "for model in structure:\n",
+    "    for chain in model:\n",
+    "        for residue in chain:\n",
+    "            if residue.resname == \"TRP\":\n",
+    "                total_TRP_residues += 1\n",
+    "            for atom in residue:\n",
+    "                if atom.element==\"C\":\n",
+    "                    total_carbon_atoms += 1 \n",
+    "\n",
+    "print(\"Total number of TRP aminoacids in the structure:\", total_TRP_residues)\n",
+    "print(\"Total number of carbon atoms in the structure:  \", total_carbon_atoms)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": false
+   },
+   "source": [
+    "We can leverage the coordinates provided by the atoms to find out the `bounding box` around the structure - by finding out the minimum and maximum value for each dimension."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 176,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Minimum coordinates for each dimension:          [121.966, 115.246, 71.133]\n",
+      "Maximum coordinates for each dimension:          [187.875, 192.4, 203.823]\n"
+     ]
+    }
+   ],
+   "source": [
+    "# We start from extreme values and expect them to go down once we encounter a `better` value.\n",
+    "min_atom_coord = [1000, 1000, 1000]\n",
+    "max_atom_coord = [-1000, -1000, -1000]\n",
+    "\n",
+    "for model in structure:\n",
+    "    for chain in model:\n",
+    "        for residue in chain:\n",
+    "            for atom in residue:\n",
+    "                coord=atom.coord\n",
+    "                for dim, val in enumerate(coord):\n",
```

Comment:
Maybe add a quick comment to explain enumerate

Reply:
Added several cells above where it's first introduced.